addons/l10n_in_edi/models/account_edi_format.py

```diff
@@ -0,0 +1,442 @@
+        json_dump = json.dumps(self._l10n_in_edi_generate_json(move))
+        return json_dump.encode()
+
+    def _check_move_configuration(self, move):
+        if self.code != "in_einvoice_1_03":
+            return super()._check_move_configuration(move)
+        error_message = []
+        error_message += self._l10n_in_validate_partner(move.partner_id)
+        error_message += self._l10n_in_validate_partner(move.company_id.partner_id, is_company=True)
+        if not move.company_id.l10n_in_edi_username or not move.company_id.sudo().l10n_in_edi_password:
+            error_message.append(_(
+                "A username and password still needs to be set for the E-invoice(IN)."
+                "\nIt needs to be added in the Settings."))
+        if not re.match("^.{1,16}$", move.name):
+            error_message.append(_("Invoice number should not be more than 16 characters"))
+        for line in move.invoice_line_ids.filtered(lambda line: not line.display_type):
+            if line.product_id:
+                if not line.product_id.l10n_in_hsn_code:
+                    error_message.append(_("HSN code is not set in product %s", line.product_id.name))
+                elif not re.match("^[0-9]+$", line.product_id.l10n_in_hsn_code):
+                    error_message.append(_(
+                        "Invalid HSN Code (%s) in product %s", line.product_id.l10n_in_hsn_code, line.product_id.name
+                    ))
+            else:
+                error_message.append(_("product is required to get HSN code"))
+        return error_message
+
+    def _l10n_in_edi_get_iap_buy_credits_message(self, company):
+        base_url = "https://iap-sandbox.odoo.com/iap/1/credit" if not company.l10n_in_edi_production_env else ""
+        url = self.env["iap.account"].get_credits_url(service_name="l10n_in_edi", base_url=base_url)
+        return """<p><b>%s</b></p><p>%s</p>""" % (
+            _("You have insufficient credits to send this document!"),
+            _("Please proceed to buy more credits <a href='%s'>here.</a>", html_escape(url)),
+        )
+
+    def _post_invoice_edi(self, invoices):
+        if self.code != "in_einvoice_1_03":
+            return super()._post_invoice_edi(invoices)
+        response = {}
+        res = {}
+        WebService = self.env["l10n.in.edi.web.service"]
+        generate_json = self._l10n_in_edi_generate_json(invoices)
+        response = WebService._generate(invoices.company_id, generate_json)
+        if response.get("error"):
+            error = response["error"]
+            error_codes = [e.get("code") for e in error]
+            if "1005" in error_codes:
+                # Invalid token eror then create new token and send generate request again.
+                # This happen when authenticate called from another odoo instance with same credentials (like. Demo/Test)
+                authenticate_response = WebService._authenticate(invoices.company_id)
```

Comment:
Having a `WebService` that doesn't handle neither the errors nor the json generation is quite useless. That's the reason why I think your extra class should not be an odoo class. It should be like a blackbox doing "magically" all the web services management and being adapted if the api evolved without being overridden nor "fixed" due to the stable policy

Reply:
okay, got it.